runner/src/server/plugins/engine/components/ComponentBase.ts

```diff
@@ -0,0 +1,58 @@
+import { Schema as JoiSchema } from "joi";
+import {
+  Data,
+  ComponentDef,
+  ContentComponentsDef,
+  InputFieldsComponentsDef,
+  StaticValues,
+} from "@xgovformbuilder/model";
+
+import { FormModel } from "../models";
+import { FormData, FormSubmissionErrors } from "../types";
+import { ViewModel } from "./types";
+
+export class ComponentBase {
+  type: ComponentDef["type"];
+  name: ComponentDef["name"];
+  title: ComponentDef["title"];
+  schema: ComponentDef["schema"];
+  options: ComponentDef["options"];
+  hint?: InputFieldsComponentsDef["hint"];
+  content?:
+    | {
+        title: string;
+        text: string;
+        condition?: any;
+      }
+    | ContentComponentsDef["content"];
+
+  model: FormModel;
+  values?: StaticValues;
+
+  formSchema?: JoiSchema;
+  stateSchema?: JoiSchema;
+
+  constructor(def: ComponentDef, model: FormModel) {
+    // component definition properties
+    this.type = def.type;
+    this.name = def.name;
+    this.title = def.title;
+    this.schema = def.schema || {};
+    this.options = def.options;
+    this.hint = "hint" in def ? def.hint : undefined;
+    this.content = "content" in def ? def.content : undefined;
```

Comment:
No, TS complains because `ComponentDef` can be any component, including those without content.

```javascript
constructor(def: ComponentDef, model: FormModel) {
```

Reply:
Just to add to this, the root cause of this issue is because ComponentBase is the base for all components,  love TS exposing design issues, ideally we could break it into the 3 bases, InputComponentBase, ListComponentBase and ContentComponentBase, with that we can also avoid the constant `prop in options` checks you've seen